docs/DeveloperGuide.adoc

```diff
@@ -142,105 +142,706 @@ image::StorageClassDiagram.png[]
+* `Model#rollback()` -- Inversely executes the latest command stored in `executedCommands` to revert the changes of the latest executed command made onto `Model`.
+* `Model#commit()` -- Executes the latest undone command stored in `inverselyExecutedCommands` to reapply the changes that were made onto `Model` by the latest undone command.
 
-image::UndoRedoState2.png[]
+Commands support the given operations to mutate the state of the `Model` and to check if they should be stored for undo/redo function:
 
-[NOTE]
-If a command fails its execution, it will not call `Model#commitAddressBook()`, so the address book state will not be saved into the `addressBookStateList`.
+* `Command#hasInverseExecution()` -- Checks if the command's execution mutates the state of the `Model`, which is used to determine if the command should be remembered by `HistoryManager`.
+* `Command#execute(Model)` -- Executes the command on the given `Model`.
+* `Command#executeInverse(Model)` -- Executes on the given `Model` such that it will undo whatever changes were made when `Command#execute(Model)` was called.
 
-Step 4. The user now decides that adding the person was a mistake, and decides to undo that action by executing the `undo` command. The `undo` command will call `Model#undoAddressBook()`, which will shift the `currentStatePointer` once to the left, pointing it to the previous address book state, and restores the address book to that state.
+Below is a class diagram between `Model`, `ModelManager`, `HistoryManager`, `CommandDeque` and `Command`.
 
-image::UndoRedoState3.png[]
+.Class Diagram for `Model`, `ModelManager`, `HistoryManager`, `CommandDeque` and `Command`
+image::HistoryManagerClassDiagram.png[]
 
-[NOTE]
-If the `currentStatePointer` is at index 0, pointing to the initial address book state, then there are no previous address book states to restore. The `undo` command uses `Model#canUndoAddressBook()` to check if this is the case. If so, it will return an error to the user rather than attempting to perform the undo.
+Undo and redo operations are executed with `UndoCommand` and `RedoCommand`. These commands store an integer value referencing the number of commands to undo or redo,
+represented by `UndoCommand#numberOfTimes` and `RedoCommand#numberOfTimes`. The Class Diagram below shows details about `UndoCommand` and `RedoCommand`.
 
-The following sequence diagram shows how the undo operation works:
+.Class Diagram for `UndoCommand`, `RedoCommand` and `Command`
+image::UndoRedoCommandClassDiagram.png[]
 
-image::UndoSequenceDiagram.png[]
+Below is a Sequence Diagram of how an `UndoCommand` executes in the program. `RedoCommand` follows a similar process.
 
-NOTE: The lifeline for `UndoCommand` should end at the destroy marker (X) but due to a limitation of PlantUML, the lifeline reaches the end of diagram.
+.Sequence Diagram for `UndoCommand`
+image::UndoCommandSequenceDiagram.png[]
```

Comment:
The getLatestCommand(), deleteLatestCommand() and addLatestCommand() section is rather confusing. We don't know which command addLatestCommand() adds. Does it add c again after deleting it or does it add the inverse of c?